os/arch/arm/src/amebasmart/amebasmart_idle.c

```diff
@@ -23,14 +23,160 @@
+
+		/* Then force the global state change */
+		ret = pm_changestate(PM_IDLE_DOMAIN, newstate);
+		if (ret < 0) {
+			/* The new state change failed, revert to the preceding state */
+			pmdbg("State change failed! Current state = %d, newstate = %d\n", oldstate, newstate);
+			newstate = oldstate;
+		} else {
+			/* Save the new state */
+			oldstate = newstate;
+		}
+		/* MCU-specific power management logic */
+		switch (newstate) {
+			case PM_NORMAL:
+				break;
+			case PM_IDLE:
+				break;
+			case PM_STANDBY:
+				break;
+			case PM_SLEEP:
+				ap_timer_helper();
+				if (up_cpu_index() == 0) {
+					/* mask sys tick interrupt*/
+					arm_arch_timer_int_mask(1);
+					up_timer_disable();
+					flags = irqsave();
+					if (tizenrt_ready_to_sleep()) {
+// Consider for dual core condition
+#ifdef CONFIG_SMP
+						/*PG flow */
+						if (pmu_get_sleep_type() == SLEEP_PG) {
+							/* CPU1 just come back from pg, so can't sleep here */
+							if (pmu_get_secondary_cpu_state(1) == CPU1_WAKE_FROM_PG) {
+								goto EXIT;
+							}
+
+							/* CPU1 is in task schedular, tell CPU1 to enter hotplug */
+							if (pmu_get_secondary_cpu_state(1) == CPU1_RUNNING) {
+								/* CPU1 may in WFI idle state. Wake it up to enter hotplug itself */
+								up_irq_enable();
+								arm_gic_raise_softirq(1, 0);
+								arm_arch_timer_int_mask(0);
+								DelayUs(100);
+								goto EXIT;
+							}
+							/* CG flow */
+						} else {
+							if (!check_wfi_state(1)) {
+								goto EXIT;
+							}
+						}
+#endif
+						// Interrupt source from BT/UART will wake cpu up, just leave expected idle time as 0
+						// Enter sleep mode for AP
+						configPRE_SLEEP_PROCESSING(xModifiableIdleTime);
+						/* When wake from pg, arm timer has been reset, so a new compare value is necessary to
+						trigger an timer interrupt */
+						if (pmu_get_sleep_type() == SLEEP_PG) {
+							up_timer_enable();
+							arm_arch_timer_set_compare(arm_arch_timer_count() + 50000);
+						}
+						arm_arch_timer_int_mask(0);
+						configPOST_SLEEP_PROCESSING(xModifiableIdleTime);
+					}
+					else {
+						/* power saving when idle*/
+						arm_arch_timer_int_mask(0);
+						__asm(" DSB");
+						__asm(" WFI");
+						__asm(" ISB");
+					}
+#ifdef CONFIG_SMP
+EXIT:
+#endif				
+					/* Re-enable interrupts and sys tick*/
+					up_irq_enable();
+				}
+				// This case is consideration for secondary core
+				else if (up_cpu_index() == 1) {
+					if (pmu_get_sleep_type() == SLEEP_PG) {
+						if (tizenrt_ready_to_sleep()) {
+							/* CPU1 will enter hotplug state. Raise a task yield to migrate its task */
+							pmu_set_secondary_cpu_state(1, CPU1_HOTPLUG);
+							// Check portYIELD();
+							// portYIELD();
+						}
+					}
+
+					flags = irqsave();
+					__asm("	DSB");
+					__asm("	WFI");
+					__asm("	ISB");
+					up_irq_enable();
+				}
+				ret = pm_changestate(PM_IDLE_DOMAIN, PM_NORMAL);
+				if (ret < 0) {
+					oldstate = PM_NORMAL;
```

Comment:
lets add a TODO comment here please
//TODO: If we are unable to reset the state back to PM_NORMAL then what to do?

Reply:
@edwakuwaku whats your opinion on this?

Reply:
I couldn't think of any scenario that will cause fail in changing state here, I suggest that we should run with application first, and check whether your concern is going to happen.

Reply:
Yes, I agree with you that unexpected things happen where there exists some corner cases we might misconsider it. Lets follow your suggestion first to enhance the effectiveness when debugging.
May I add todo comment and the PM log from some other discussion thread together in a new commit?

Reply:
I understand it can be very irritating to rebase and make changes to lower-stacked commits sometimes but modularity of commits is very important for us for easy look-back and reference in the future.
Hence we requested you to separate non-PM changes in a separate PR and for the same reason, I request you to add all debug statement and TODO comments in the same commits itself please.

Reply:
Sure no problem~